src/brain/operating_unit_recorder.cpp

```diff
@@ -0,0 +1,240 @@
+    plan_features_.insert(features.begin(), features.end());
+  }
+
+  for (auto key : plan->GetRightHashKeys()) {
+    auto features = ExtractFeaturesFromExpression(key);
+    plan_features_.insert(features.begin(), features.end());
+  }
+}
+
+void OperatingUnitRecorder::Visit(const planner::NestedLoopJoinPlanNode *plan) {
+  VisitAbstractJoinPlanNode(plan);
+
+  for (auto key : plan->GetLeftKeys()) {
+    auto features = ExtractFeaturesFromExpression(key);
+    plan_features_.insert(features.begin(), features.end());
+  }
+
+  for (auto key : plan->GetRightKeys()) {
+    auto features = ExtractFeaturesFromExpression(key);
+    plan_features_.insert(features.begin(), features.end());
+  }
+}
+
+void OperatingUnitRecorder::Visit(const planner::LimitPlanNode *plan) { VisitAbstractPlanNode(plan); }
+
+void OperatingUnitRecorder::Visit(const planner::OrderByPlanNode *plan) { VisitAbstractPlanNode(plan); }
+
+void OperatingUnitRecorder::Visit(const planner::ProjectionPlanNode *plan) { VisitAbstractPlanNode(plan); }
+
+void OperatingUnitRecorder::Visit(const planner::AggregatePlanNode *plan) { VisitAbstractPlanNode(plan); }
+
+OperatingUnitFeatureVector OperatingUnitRecorder::RecordTranslators(
+    const std::vector<std::unique_ptr<execution::compiler::OperatorTranslator>> &translators) {
+  // Note that OperatorTranslators are roughly 1:1 with a plan node
+  // As such, just emplace directly into feature vector
+  std::vector<OperatingUnitFeature> results{};
+  std::unordered_set<const planner::AbstractPlanNode *> plan_nodes;
+  for (const auto &translator : translators) {
+    // TODO(wz2): Populate actual num_rows/cardinality after #759
+    auto feature_type = translator->GetFeatureType();
+    auto num_rows = 0;
+    auto cardinality = 0.0;
+    if (feature_type != OperatingUnitFeatureType::INVALID && feature_type != OperatingUnitFeatureType::OUTPUT) {
+      plan_nodes.insert(translator->Op());
+      results.emplace_back(feature_type, num_rows, cardinality);
+    }
+  }
+
+  std::unordered_map<OperatingUnitFeatureType, OperatingUnitFeature> features;
+  for (auto *plan : plan_nodes) {
+    // Consolidate the features based on OutputSchema
+    plan_features_ = {};
+    plan->Accept(common::ManagedPointer<planner::PlanVisitor>(this));
+
+    for (auto feature : plan_features_) {
+      // TODO(wz2): Get these from cost model/plan?
+      size_t num_rows = 0;
+      auto cardinality = 0.0;
+      auto itr = features.find(feature);
+      if (itr == features.end()) {
+        features.insert(std::make_pair(feature, OperatingUnitFeature(feature, num_rows, cardinality)));
+      } else {
+        // For now, we just record the max cardinality/max estimated rows
+        // for cases of duplicate fatures (i.e ADD). In the future, may
+        // want to change the "de-duplication" approach (i.e maybe store
+        // all or store the data associated with the plan that emits the
+        // most amount of tuples)
```

Comment:
For the duplicated features, we can just add all of them up for now (the cardinalities and the number of rows). Think about it as we're trying to record how much work in total this pipeline is going to do.